packages/database/src/migrations/20210824054133-FixAnalyticTableWeekPeriodIncorrectYear-modifies-schema.js

```diff
@@ -0,0 +1,116 @@
+        to_char(survey_response.data_time, ''YYYYMM'') as "month_period",
+        to_char(survey_response.data_time, ''YYYY'') as "year_period",
+        survey_response.data_time as "date"
+      FROM
+        survey_response
+      INNER JOIN
+        answer ON answer.survey_response_id = survey_response.id
+      INNER JOIN
+        entity ON entity.id = survey_response.entity_id
+      INNER JOIN
+        survey ON survey.id = survey_response.survey_id
+      INNER JOIN
+        question ON question.id = answer.question_id
+      INNER JOIN
+        data_source ON data_source.id = question.data_source_id
+      WHERE data_source.service_type = ''tupaia'' AND survey_response.outdated IS FALSE';
+  
+  begin
+    RAISE NOTICE 'Creating Materialized View Logs...';
+  
+    FOREACH source_table IN ARRAY source_tables_array LOOP
+      IF (SELECT NOT EXISTS (
+        SELECT FROM pg_tables
+        WHERE schemaname = 'public'
+        AND tablename   = 'log$_' || source_table
+      ))
+      THEN
+        EXECUTE 'ALTER TABLE ' || source_table || ' DISABLE TRIGGER ' || source_table || '_trigger';
+        tStartTime := clock_timestamp();
+        PERFORM mv$createMaterializedViewlog(source_table, 'public');
+        RAISE NOTICE 'Created Materialized View Log for % table, took %', source_table, clock_timestamp() - tStartTime;
+        EXECUTE 'ALTER TABLE ' || source_table || ' ENABLE TRIGGER ' || source_table || '_trigger';
+      ELSE
+        RAISE NOTICE 'Materialized View Log for % table already exists, skipping', source_table;
+      END IF;
+    END LOOP;
+  
+  
+    RAISE NOTICE 'Creating analytics materialized view...';
+    IF (SELECT NOT EXISTS (
+      SELECT FROM pg_tables
+      WHERE schemaname = 'public'
+      AND tablename   = 'analytics'
+    ))
+    THEN
+      tStartTime := clock_timestamp();
+      PERFORM mv$createMaterializedView(
+          pViewName           => 'analytics',
+          pSelectStatement    =>  pSqlStatement,
+          pOwner              => 'public',
+          pFastRefresh        =>  TRUE
+      );
+      RAISE NOTICE 'Created analytics table, took %', clock_timestamp() - tStartTime;
+    ELSE
+      IF (force)
+      THEN
+      RAISE NOTICE 'Force rebuilding analytics table';
+        tStartTime := clock_timestamp();
+        PERFORM mv$createMaterializedView(
+            pViewName           => 'analytics_tmp',
+            pSelectStatement    =>  pSqlStatement,
+            pOwner              => 'public',
+            pFastRefresh        =>  TRUE
+        );
+        RAISE NOTICE 'Created analytics table, took %', clock_timestamp() - tStartTime;
+        PERFORM mv$removeMaterializedView('analytics', 'public');
+        PERFORM mv$renameMaterializedView('analytics_tmp', 'analytics', 'public');
+      ELSE
+        RAISE NOTICE 'Analytics Materialized View already exists, skipping';
+      END IF;
+    END IF;
+  end $$ LANGUAGE plpgsql;
+`);
```

Comment:
So just a heads up, there's one more thing you gotta do here. You'll need to actually rebuild the analytics table now that we've changed the schema for it. This can be done by calling:
```
await db.runSql('SELECT build_analytics_table(true);');
await db.runSql('SELECT create_analytics_table_indexes();');
```

However, this migration will then take several minutes to run... There is a slightly hacky option we could use if we wish to avoid running the rebuild of the analytics table here... we could update the existing migration? That would work, since the first migration hasn't made it to prod as yet so it's still being run each night... quite hacky but might be a good idea..?

Reply:
Oh that is a good point! If that is not in prod, we can just change the existing migration👏

Reply:
Also noted the steps you shared to rebuild the analytics table. Thanks @rohan-bes .